### PR TITLE
Ensemble configs

### PR DESCRIPTION
**Pull request by rvashurin**

### CODE REVIEW

File: scripts/polygraph_eval [hunk 1]
```diff
@@ -15,7 +15,7 @@ log = logging.getLogger()
 
 from lm_polygraph.utils.manager import UEManager
 from lm_polygraph.utils.dataset import Dataset
-from lm_polygraph.utils.model import WhiteboxModel
```
Comment (rvashurin) [inline scripts/polygraph_eval:18]: Создание ансамбля теперь в model.py

File: scripts/polygraph_eval [hunk 2]
```diff
@@ -66,19 +66,37 @@ def main(args):
...
+                                             device='cpu',
+                                             seed=args.seed[0],
+                                             ensembling_mode=args.ensembling_mode,
+                                             mc_seeds=args.mc_seeds,
+                                             dropout_rate=float(args.dropout_rate),
+                                             **cache_kwargs,
+                                             **model_kwargs
+                                             )
+        else:
+            ensemble_model = None
```
Comment (rvashurin) [inline scripts/polygraph_eval:98]: Если в конфиге прописан ансамбль - создаем ансамбль и кладем его в переменную. Если не прописан - в переменной None, который потом в менеджере проверится и не станет запускать ансамблевые расчеты.

File: scripts/polygraph_eval [hunk 3]
```diff
@@ -146,14 +164,14 @@ def main(args):
...
-                args.background_train_dataset_text_column, 
+                args.background_train_dataset_text_column,
                 args.background_train_dataset_label_column,
                 batch_size=args.batch_size,
                 data_files=args.background_train_dataset_data_files,
                 split="train",
                 size=100_000,
                 load_from_disk=args.background_load_from_disk,
-                **data_kwargs
+                **cache_kwargs
```
Comment (rvashurin) [inline scripts/polygraph_eval:174]: Просто старый баг.

File: scripts/polygraph_eval [hunk 4]
```diff
@@ -196,6 +214,7 @@ def main(args):
             ignore_exceptions=args.ignore_exceptions,
             background_train_data=background_train_dataset,
             max_new_tokens=args.max_new_tokens,
+            ensemble_model=ensemble_model
```
Comment (rvashurin) [inline scripts/polygraph_eval:217]: Передаем ансамбль в менеджер. Либо модель, либо None.

File: scripts/polygraph_eval [hunk 5]
```diff
@@ -270,6 +289,43 @@ def get_ue_methods(args, model):
             Eccentricity(similarity_score="Jaccard_score"),
             SemanticEntropy(),
         ]
+
+    if args.use_ens_ue:
```
Comment (rvashurin) [inline scripts/polygraph_eval:293]: Если юзаем ансамблевые меры - добавляем эстиматоры в список.

File: src/lm_polygraph/app/parsers.py [hunk 6]
```diff
@@ -174,14 +173,6 @@ def parse_ensemble(path: str, device: str = 'cpu') -> Tuple[WhiteboxModel, White
     devices = [device] * (len(model_paths) - 1)
 
     model = WhiteboxModel.from_pretrained(model_paths[0])
-    ensemble_model = WhiteboxModel.from_pretrained(model_paths[0])
```
Comment (rvashurin) [inline src/lm_polygraph/app/parsers.py:177]: Теперь создание ансамбля живет в model.py, тут этот код не нужен.

File: src/lm_polygraph/utils/ensemble_utils/dropout.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    children = module.children()
+    for c_mod in children:
+        if type(c_mod).__name__ == "Dropout":
+            method = getattr(ConsistentDropout, 'identity')
+            setattr(c_mod, 'forward', types.MethodType(method, c_mod))
+        else:
+            replace_with_identity(c_mod)
+
+
+def replace_dropout(model_name, module, p=0.1, share_across_tokens=True):
```
Comment (rvashurin) [inline src/lm_polygraph/utils/ensemble_utils/dropout.py:58]: Тут сложно, идея в том, чтобы заменить дропаут слои в модели на свои (`ConsistentDropout`) который будут применять одинаковую маску на всех шагах декодирования и для всех элементов в батче.

BARTы при этом вместо модуля dropout юзают torch.nn.functional.dropout поэтому с ними особый разговор.

**Review by cant-access-rediska0123 — APPROVED** [review]
